crates/cext/src/circuit.rs

```diff
@@ -409,6 +413,115 @@ pub struct OpCounts {
     len: usize,
 }
 
+#[inline]
+fn conjugate(matrix: &ArrayView2<Complex64>) -> Array2<Complex64> {
+    Array2::from_shape_fn((matrix.nrows(), matrix.ncols()), |(i, j)| {
+        matrix[(j, i)].conj()
+    })
+}
+
+/// Check an [ArrayType] represents a unitary matrix. Uses an element-wise check; if
+/// any element in ``conjugate(matrix) * matrix`` differs from the identity by more than ``tol``
+/// (in magnitude), the matrix is not considered unitary.
+fn is_unitary(matrix: &ArrayType, tol: f64) -> bool {
+    let not_unitary = match matrix {
+        ArrayType::OneQ(mat) => (mat.adjoint() * mat - Matrix2::identity())
+            .iter()
+            .any(|val| val.abs() > tol),
+        ArrayType::TwoQ(mat) => (mat.adjoint() * mat - Matrix4::identity())
+            .iter()
+            .any(|val| val.abs() > tol),
+        ArrayType::NDArray(mat) => {
+            let product = mat.dot(&conjugate(&mat.view()));
+            let n = mat.nrows();
+            let one = Complex64::new(1.0, 0.0);
+            product.iter().enumerate().any(|(index, value)| {
+                let col_idx = index % n;
+                let row_idx = index / n;
+                if col_idx == row_idx {
+                    (value - one).abs() > tol
+                } else {
+                    value.abs() > tol
+                }
+            })
+        }
+    };
+    !not_unitary // using double negation to use ``any`` (faster) instead of ``all``
+}
+
+/// @ingroup QkCircuit
+/// Append an arbitrary unitary matrix to the circuit.
+///
+/// @param circuit A pointer to the circuit to append the unitary to.
+/// @param matrix A pointer to the ``QkComplex64`` array representing the unitary matrix.
+///     This must be a row-major, unitary matrix of dimension ``2 ^ num_qubits x 2 ^ num_qubits``.
+///     More explicitly: the ``(i, j)``-th element is given by ``matrix[i * 2^n + j]``.
```

Comment:
We should mention that matrix is copied here so the caller knows they retain ownership of the input array and it's safe to free.